File: frontend/src/services/items.ts
import api from "./api";
import type { Item, PaginatedResponse } from "../types";

export interface ItemFilters {
  item_type?: string;
  category?: string;
  status?: string;
  container_id?: string;
  low_stock?: boolean;
  search?: string;
  page?: number;
  page_size?: number;
  sort_by?: string;
  sort_order?: "asc" | "desc";
}

export async function getItems(params?: ItemFilters) {
  const { data } = await api.get<PaginatedResponse<Item>>("/items", { params });
  return data;
}

export async function getItem(id: string) {
  const { data } = await api.get<Item>(`/items/${id}`);
  return data;
}

export async function createItem(item: Record<string, unknown>) {
  const { data } = await api.post<Item>("/items", item);
  return data;
}

export async function updateItem(id: string, item: Partial<Item>) {
  const { data } = await api.patch<Item>(`/items/${id}`, item);
  return data;
}

export async function deleteItem(id: string) {
  await api.delete(`/items/${id}`);
}

export async function adjustQuantity(id: string, delta: number, note?: string) {
  const { data } = await api.post<Item>(`/items/${id}/adjust`, { delta, note });
  return data;
}

export async function changeStatus(id: string, status: string, assigned_to?: string) {
  const { data } = await api.patch<Item>(`/items/${id}/status`, { status, assigned_to });
  return data;
}

export async function moveItem(id: string, container_id?: string, parent_item_id?: string) {
  const { data } = await api.patch<Item>(`/items/${id}/move`, { container_id, parent_item_id });
  return data;
}

export async function uploadItemImage(itemId: string, file: File) {
  const formData = new FormData();
  formData.append("file", file);
  const { data } = await api.post<{ image_url: string }>(`/uploads/image/${itemId}`, formData, {
    headers: { "Content-Type": "multipart/form-data" },
  });
  return data;
}

export async function deleteItemImage(itemId: string) {
  await api.delete(`/uploads/image/${itemId}`);
}

File: backend/app/api/v1/endpoints/items.py
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.v1.deps import get_current_user
from app.core.database import get_db
from app.models.user import User
from app.repositories.item_repository import ItemRepository
from app.schemas.item import (
    AdjustPayload,
    ItemCreate,
    ItemResponse,
    ItemUpdate,
    MovePayload,
    PaginatedItems,
    StatusPayload,
)
from app.services.inventory_service import InventoryService

router = APIRouter()


@router.get("", response_model=PaginatedItems)
async def list_items(
    item_type: str | None = None,
    category: str | None = None,
    status: str | None = None,
    container_id: UUID | None = None,
    low_stock: bool = False,
    search: str | None = None,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    sort_by: str = "updated_at",
    sort_order: str = "desc",
    db: AsyncSession = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    repo = ItemRepository(db)
    items, total = await repo.list_items(
        item_type=item_type,
        category=category,
        status=status,
        container_id=container_id,
        low_stock=low_stock,
        search=search,
        page=page,
        page_size=page_size,
        sort_by=sort_by,
        sort_order=sort_order,
    )
    return PaginatedItems(items=items, total=total, page=page, page_size=page_size)


@router.post("", response_model=ItemResponse, status_code=201)
async def create_item(
    data: ItemCreate,
    db: AsyncSession = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    svc = InventoryService(db)
    return await svc.create_item(data)


@router.get("/{item_id}", response_model=ItemResponse)
async def get_item(
    item_id: UUID,
    db: AsyncSession = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    repo = ItemRepository(db)
    item = await repo.get_by_id(item_id)
    if not item:
        from fastapi import HTTPException
        raise HTTPException(status_code=404, detail="Item not found")
    return item


@router.patch("/{item_id}", response_model=ItemResponse)
async def update_item(
    item_id: UUID,
    data: ItemUpdate,
    db: AsyncSession = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    svc = InventoryService(db)
    return await svc.update_item(item_id, data.model_dump(exclude_unset=True))


@router.delete("/{item_id}", status_code=204)
async def delete_item(
    item_id: UUID,
    db: AsyncSession = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    svc = InventoryService(db)
    await svc.delete_item(item_id)


@router.post("/{item_id}/adjust", response_model=ItemResponse)
async def adjust_quantity(
    item_id: UUID,
    payload: AdjustPayload,
    db: AsyncSession = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    svc = InventoryService(db)
    return await svc.adjust_quantity(item_id, payload)


@router.patch("/{item_id}/status", response_model=ItemResponse)
async def change_status(
    item_id: UUID,
    payload: StatusPayload,
    db: AsyncSession = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    svc = InventoryService(db)
    return await svc.change_status(item_id, payload)


@router.patch("/{item_id}/move", response_model=ItemResponse)
async def move_item(
    item_id: UUID,
    payload: MovePayload,
    db: AsyncSession = Depends(get_db),
    _user: User = Depends(get_current_user),
):
    svc = InventoryService(db)
    return await svc.move_item(item_id, payload)
